### ISSUE TITLE
Find out why SimpleBarrier seems to cause process lock up

### ISSUE THREAD

**Issue by hhyyrylainen** [Revolutionary-Games/Thrive#4996]
It seems to happen even though the threads don't even get stuck in anything related to it, but suddenly when switching to the default Barrier class that lock up goes away. I can't figure this out now so I'm opening an issue for this instead.

**Comment by hhyyrylainen** [comment]
This can be visited sometime in the future, but for now I don't see how this could be fixed by *us*.

**Comment by xfractalino** [comment]
It looks like a race condition. I caught the bug in the following test:

```
using System;
using System.Threading;
using System.Threading.Tasks;

public class Program
{
    static SimpleBarrier barrier = new(2);
    static long iterations = 0;

    public static void Test()
    {
        Console.WriteLine("Starting Race Condition Test...");

        var t1 = Task.Run(() =>
        {
            while (true)
            {
                // this thread quickly goes back to wait
                barrier.SignalAndWait();
            }
        });

        var t2 = Task.Run(() =>
        {
            while (true)
            {
                barrier.SignalAndWait();

                // this takes a bit longer to write the cycles
                long current = Interlocked.Increment(ref iterations);
                Console.Write($"\rCycles: {current}");
            }
        });

        Task.WaitAll(t1, t2);
    }
}
```

The program runs for a few hundred cycles, before deadlocking. During the deadlock, `readCount` is stuck on 1, and one thread is stuck waiting in the `while (readCount != threadCount)` at line 60 in `SimpleBarrier`. The second one, though, is slower, and is still in  `while (readCount != 0)` on line 113 because the first one has already incremented `readCount` before the second even exited the `SignalAndWait`. So it looks like a scheduling issue that causes this race condition.

**Comment by hhyyrylainen** [comment]
Ah, so the problem is that the barrier would need to have one more step in it that prevents a thread from leaving before all threads are at the leave step? I must have overlooked that in the design.

**Comment by xfractalino** [comment]
Yes, I guess. Just out of curiosity: why implement a custom barrier?

**Comment by hhyyrylainen** [comment]
Have you looked at the default barrier implementation? It has an absolute ton of stuff in it so I reasoned that it must be slower than a minimum viable set of barrier features.

**Comment by xfractalino** [comment]
Yes, I have. Valid concern
